[FR] Add copilot instructions to catch the gotchas (#5733)
elastic/detection-rules: .github/instructions/rules.instructions.md

---
applyTo:
  - rules/**/*.toml
  - rules_building_block/**/*.toml
---

You are a **detection engineer**. Your task is to **review pull requests** (PRs) to improve the quality of detection rules in the `detection-rules` repository. Take your time, think carefully, and perform a thorough review before writing any suggestions.

This repository supports multiple query languages (**KQL**, **EQL**, **ES|QL**) and multiple rule types (**query**, **eql**, **esql**, **threshold**, **new_terms**, **threat_match**, **machine_learning**). Each has distinct fields, constraints, and performance implications. Review accordingly.

There are different types of pull requests:

- **New rule:** The PR name often starts with `[New Rule]`
- **Tuning an existing rule:** The PR name often starts with `[Tuning]`
- **Deprecating a rule:** The PR name often starts with `[Deprecation]`
- **New/tuning hunt:** The PR name often starts with `[Hunt]`
- **Mixed PRs:** May combine tuning, new rules, or deprecations

> For each rule `.toml` file in the PR, review the **metadata**, **rule fields**, and **query** sections using the following guidance:

Reviewer: ```suggestion
For each rule `.toml` file in the PR, review the **metadata** and **rule** sections paying close attention to the **query** value in the **rule** section using the following guidance:
```

Nit this might be more clear for the AI, not sure.

- Assess and suggest a better `rule.name` if the existing name does not accurately reflect the rule's query logic and detection scope.
- Assess and suggest improvements to `rule.description` — limit to two or three sentences that clearly summarize the query logic and detection scope.
- `min_stack_version` should support the widest stack versions unless the rule uses features exclusive to a newer stack version.
- If `min_stack_comments` is set, verify it explains why `min_stack_version` is restricted.
- Report any inconsistencies in the **MITRE ATT&CK** mappings: missing relevant techniques, unrelated techniques, or missing subtechniques where applicable. When possible, suggest the most accurate and up-to-date technique/subtechnique mappings based on the query logic.
- Review the **references** section and report inconsistencies between referenced content and the rule's description or logic.

</Metadata>

---

### <Rule Fields>

#### Common Fields (All Rule Types)

- `risk_score` should align with `severity`:
  - `low` → typically 0–21
  - `medium` → typically 22–47
  - `high` → typically 48–73

>   - `critical` → typically 74–100

Reviewer 2: ```suggestion
  - `low` → 21
  - `medium` → 47
  - `high` → 73
  - `critical` → 100
```
I don't think there are any cases where we use values outside of these

- `from` and `interval` should not create gaps. The lookback window (`from`) must cover at least the `interval` period.
- `timestamp_override` should be set to `"event.ingested"` for most rules to avoid ingestion delay issues.
- `max_signals` defaults to 100. Only override with justification.

#### Investigation Guide (`note` field)

- If present, the note should include actionable triage steps, false positive guidance, and response/remediation steps.
- The `setup` field should include necessary steps to configure the integration or data source.
- Check for typos in the investigation guide content.

#### Building Block Rules

- Must be placed in the `rules_building_block/` directory.
- Should have lower `risk_score` and `severity` than the correlation rules that consume them.

#### Alert Suppression

- Verify the suppression `group_by` fields are meaningful for deduplication.

#### Threshold Rules

- Verify the `threshold.value` is meaningful and not too low (noisy) or too high (misses detections).
- When `threshold.cardinality` is used, verify the cardinality field and value make sense for the detection logic.

#### New Terms Rules

> - `history_window_start` should be appropriate for the detection context (typically 7–14 days).

Reviewer 3: ```suggestion
- `history_window_start` should be appropriate for the detection context (typically 5–14 days) longer values may impact performance.
```

- Verify the `new_terms_fields` combination makes semantic sense — detecting "first seen" on arbitrary field combinations can produce excessive noise.

> 

Reviewer 4: ```suggestion
- Verify the `new_terms_fields` combination makes semantic sense — detecting "first seen" on arbitrary field combinations can produce excessive noise.
- Assess whether it is truly necessary to leverage multiple `new_terms_fields` keys, as each newly added key negatively impacts performance.
```

#### Threat Match Rules

- `threat_indicator_path` should be set (default: `threat.indicator`).
- `from` and `interval` for threat match rules are typically wider (e.g., `from: "now-65m"`, `interval: "1h"`) due to indicator matching overhead.

#### Machine Learning Rules

- `setup` should document the required ML job installation steps.

</Rule Fields>

---

### <Query — All Languages>

- Review for typos in known system file names (e.g., `WmiPrvS.exe` instead of `WmiPrvSe.exe`).
- Ensure the **query logic aligns with the rule description** (e.g., the description says "Detect Certutil abuse," but the query looks for `svchost.exe`).
- Verify there are no duplicate entries in the query (e.g., same exclusion listed twice).

> - For production rule tuning PRs, ensure changes are limited to excluding false positives and do not extend the detection scope while the rule remains in production maturity.

Reviewer 3: ```suggestion
- For rule tuning PRs, ensure changes are limited to excluding false positives and do not extend the detection scope while the rule remains in production maturity.
```

Reviewer 4: I think this comment is mostly for ER. We should remove the entire comment, as in DR, I think we can just increase the scope if it seems necessary.
